vowpalwabbit/cb_sample.cc

```diff
@@ -8,82 +8,91 @@ using namespace LEARNER;
+        {
+          chosen_action = static_cast<uint32_t>(i);
+          break;
+        }
       }
     }
-  }
-  else
-  {
-    bool tag_provided_seed = false;
-    uint64_t seed = data.seed_state;
-    if (examples[0]->tag.size() > 0)
+    else
     {
-      const std::string SEED_IDENTIFIER = "seed=";
-      if (strncmp(examples[0]->tag.begin(), SEED_IDENTIFIER.c_str(), SEED_IDENTIFIER.size()) == 0 &&
-          examples[0]->tag.size() > SEED_IDENTIFIER.size())
+      bool tag_provided_seed = false;
+      uint64_t seed = _seed_state;
+      if (!examples[0]->tag.empty())
+      {
+        const std::string SEED_IDENTIFIER = "seed=";
+        if (strncmp(examples[0]->tag.begin(), SEED_IDENTIFIER.c_str(), SEED_IDENTIFIER.size()) == 0 &&
+            examples[0]->tag.size() > SEED_IDENTIFIER.size())
+        {
+          substring tag_seed{examples[0]->tag.begin() + 5, examples[0]->tag.begin() + examples[0]->tag.size()};
+          seed = uniform_hash(tag_seed.begin, substring_len(tag_seed), 0);
+          tag_provided_seed = true;
+        }
+      }
+
+      // Sampling is done after the base learner has generated a pdf.
+      auto result = exploration::sample_after_normalizing(
+          seed, ACTION_SCORE::begin_scores(action_scores), ACTION_SCORE::end_scores(action_scores), chosen_action);
+      assert(result == S_EXPLORATION_OK);
+      _UNUSED(result);
+
+      // Update the seed state in place if it was used for this example.
+      if (!tag_provided_seed)
       {
-        substring tag_seed{examples[0]->tag.begin() + 5, examples[0]->tag.begin() + examples[0]->tag.size()};
-        seed = uniform_hash(tag_seed.begin, substring_len(tag_seed), 0);
-        tag_provided_seed = true;
+        merand48(_seed_state);
       }
     }
 
-    // Sampling is done after the base learner has generated a pdf.
-    auto result = exploration::sample_after_normalizing(
-        seed, ACTION_SCORE::begin_scores(action_scores), ACTION_SCORE::end_scores(action_scores), chosen_action);
+    auto result = exploration::swap_chosen(action_scores.begin(), action_scores.end(), chosen_action);
     assert(result == S_EXPLORATION_OK);
     _UNUSED(result);
-
-    // Update the seed state in place if it was used for this example.
-    if (!tag_provided_seed)
-    {
-      merand48(data.seed_state);
-    }
   }
 
-  auto result = exploration::swap_chosen(action_scores.begin(), action_scores.end(), chosen_action);
-  assert(result == S_EXPLORATION_OK);
-  _UNUSED(result);
+ private:
+  uint64_t _seed_state;
+};
+}  // namespace VW
+
+template <bool is_learn>
+void learn_or_predict(cb_sample_data &data, multi_learner &base, multi_ex &examples)
+{
+  data.learn_or_predict<is_learn>(base, examples);
 }
 
-base_learner* cb_sample_setup(options_i& options, vw& all)
+base_learner *cb_sample_setup(options_i &options, vw &all)
 {
-  auto data = scoped_calloc_or_throw<cb_sample_data>();
   bool cb_sample_option = false;
```

Comment:
I don't see where this is set to true?  Does this setup function always return nullptr?

Reply:
This is set to true by the option parser on line 90